src/storage/inverted_index_sparse_ann.rs

```diff
@@ -0,0 +1,206 @@
+use arcshift::ArcShift;
+use core::array::from_fn;
+use std::path::Path;
+
+use std::sync::Arc;
+
+use crate::models::{
+    buffered_io::BufferManagerFactory,
+    cache_loader::NodeRegistry,
+    lazy_load::{LazyItem, LazyItemArray, LazyItemVec},
+    types::SparseVector,
+};
+
+// TODO: Add more powers for larger jumps
+// TODO: Or switch to dynamic calculation of power of max power of 4
+const POWERS_OF_4: [u32; 8] = [1, 4, 16, 64, 256, 1024, 4096, 16384];
+
+/// Returns the largest power of 4 that is less than or equal to `n`.
+/// Iteratively multiplies by 4 until the result exceeds `n`.
+pub fn largest_power_of_4_below(n: u32) -> (usize, u32) {
+    assert_ne!(n, 0, "Cannot find largest power of 4 below 0");
+    POWERS_OF_4
+        .into_iter()
+        .enumerate()
+        .rev()
+        .find(|&(_, pow4)| pow4 <= n)
+        .unwrap()
+}
+
+/// Calculates the path from `current_dim_index` to `target_dim_index`.
+/// Decomposes the difference into powers of 4 and returns the indices.
+fn calculate_path(target_dim_index: u32, current_dim_index: u32) -> Vec<usize> {
+    let mut path = Vec::new();
+    let mut remaining = target_dim_index - current_dim_index;
+
+    while remaining > 0 {
+        let (child_index, pow_4) = largest_power_of_4_below(remaining);
+        path.push(child_index);
+        remaining -= pow_4;
+    }
+
+    path
+}
+
+/// [InvertedIndexSparseAnnNode] (earlier InvertedIndexItem) is a node in InvertedIndexSparseAnn structure
+/// data in InvertedIndexSparseAnnNode holds list of Vec_Ids corresponding to the quantized u8 value (which is the index of array)
+#[derive(Clone)]
+pub struct InvertedIndexSparseAnnNode {
+    pub dim_index: u32,
+    pub implicit: bool,
+    pub data: Arc<[LazyItemVec<u32>; 63]>,
```

Comment:
Let's benchmark this. I have a hunch this structure will cause a lot of memory pressure when inserting large amounts of data. It can be compared with Arc<Mutex<Vec>> implementation, ignoring serialization for now.

It should eventually be replaced with a structure that is purpose designed for concurrent appends and reads and lazy serde.